core/src/main/java/org/infinispan/functional/impl/MetaParams.java

```diff
@@ -150,62 +240,88 @@ public String toString() {
+      return new MetaParams(EMPTY_ARRAY, 0);
    }
 
    private static MetaParam[] filterDuplicates(MetaParam... metas) {
       Map<Class<?>, MetaParam<?>> all = new HashMap<>();
-      for (MetaParam meta : metas)
+      for (MetaParam meta : metas) {
+         if (meta == null) continue;
          all.put(meta.getClass(), meta);
+      }
 
       return all.values().toArray(new MetaParam[all.size()]);
    }
 
    void merge(MetaParams other) {
       Map<Class<?>, MetaParam<?>> all = new HashMap<>();
       // Add other's metadata first, because we don't want to override those already present
-      for (MetaParam meta : other.metas)
+      for (MetaParam meta : other.metas) {
+         if (meta == null) continue;
          all.put(meta.getClass(), meta);
-      for (MetaParam meta : metas)
+      }
+      for (MetaParam meta : metas) {
+         if (meta == null) continue;
          all.put(meta.getClass(), meta);
+      }
 
 
       metas = all.values().toArray(new MetaParam[all.size()]);
    }
 
-   public static final class Externalizer extends AbstractExternalizer<MetaParams> {
-      @Override
-      public void writeObject(ObjectOutput oo, MetaParams o) throws IOException {
-         oo.writeInt(o.metas.length);
-         for (Object meta : o.metas) oo.writeObject(meta);
-      }
+   @Override
+   public Iterator<MetaParam<?>> iterator() {
+      return new It();
+   }
 
-      @Override
-      public MetaParams readObject(ObjectInput input) throws IOException, ClassNotFoundException {
-         int length = input.readInt();
-         MetaParam[] metas = new MetaParam[length];
-         for (int i = 0; i < length; i++)
-            metas[i] = (MetaParam) input.readObject();
+   public static MetaParams readFrom(ObjectInput input) throws IOException, ClassNotFoundException {
+      int length = input.readInt();
+      MetaParam[] metas = new MetaParam[length];
+      for (int i = 0; i < length; i++)
+         metas[i] = (MetaParam) input.readObject();
 
-         return MetaParams.of(metas);
+      return new MetaParams(metas, metas.length);
+   }
+
+   public static void writeTo(ObjectOutput output, MetaParams params) throws IOException {
+      output.writeInt(params.size());
+      for (MetaParam meta : params) output.writeObject(meta);
+   }
+
+   private class It implements Iterator<MetaParam<?>> {
+      private int i = 0;
+
+      public It() {
+         skipNulls();
+      }
+
+      private void skipNulls() {
+         while (i < metas.length && metas[i] == null) ++i;
       }
 
       @Override
-      public Set<Class<? extends MetaParams>> getTypeClasses() {
-         return Util.<Class<? extends MetaParams>>asSet(MetaParams.class);
+      public boolean hasNext() {
+         skipNulls();
+         return i < metas.length;
       }
 
       @Override
-      public Integer getId() {
-         return Ids.META_PARAMS;
+      public MetaParam<?> next() {
+         try {
+            return metas[i++];
```

Comment:
One can call `it.next().next()` without any `hasNext()` in between, so the `skipNulls()` call should be here.